Add checks for XPath for AvoidUsingInternalURL

yutingc opened the pull request.

yutingc commented:
AppVeyor is complaining about "The type or namespace name 'TypeDefinitionAst' could not be found" issue again.. but we already merged all the changes to BugFixes?

raghushantha commented:
XPath branch does not contain the appveyor.yml file. Did you fork this branch off of BugFixes?


Hence it is building on 2012R2 image with no WMF

yutingc commented:
Yes I think so...

raghushantha commented:
Can you check if all these xpaths are covered?
https://msdn.microsoft.com/en-us/library/ms256086(v=vs.110).aspx

yutingc commented:
I didn't check the syntax of XPath itself since there are a lot of them but only check if the scripts have functions "SelectSingleNode" and "SelectNodes"

raghushantha commented:
ah. ok. Changes look good.

yutingc commented:
Added more checks for any methods of XPathNavigator that takes XPathExpression as parameter.

raghushantha commented:
IS this rule deterministic? What if the script calls into an helper to process the xpath?

yutingc commented:
What do you mean by "calls into a helper"? This rule only checks the most direct methods that takes in XPathExpression. It is not 100% deterministic but it can detect most of the XPath usage I think....

raghushantha commented:
if the xpath is processed by a function in an other module (imported here), then we will display a warning.

quoctruong commented on Rules/AvoidUsingInternalURLs.cs:
```diff
@@ -44,6 +44,30 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
                 {
+                    //Check if XPath is used. If XPath is used, then we don't throw warnings.
+                    Ast parentAst = expressionAst.Parent;
+                    IEnumerable<Ast> invocation = parentAst.FindAll(test => test is InvokeMemberExpressionAst, true);
```
Should we just check that the parentAst is a InvokeMemberExpressionAst? I don't think there would be more than 1 InvokeMemberExpressionAst that we will be looking at?

yutingc commented:
I changed the check to only check one parent. Also change the severity to information. Updated the test information in Invoke-ScriptAnalyzer.Tests and GetScriptAnalyzerRules.Tests.

For the question Raghu addressed above, we will consider adding "Type Inference" feature in the future.

raghushantha commented:
pls merge the fixes..

quoctruong commented:
Looks good to me.

yutingc commented:
Thanks!
